### PR TITLE
JAMES-2292 GET /mailQueues/{mailQueueName}/messages endpoint

### PR DESCRIPTION
**Pull request by aduprat**
Only four last commits:

* JAMES-2292 Introduce MailQueueItemDTO c89bb79
* JAMES-2292 Introduce GET mail queue messages endpoint f806dca
* JAMES-2292 Filter delayed messages c8930d3
* JAMES-2292 Limit the number of messages 1ce27db

### CODE REVIEW

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+        private Builder() {
+        }
+
+        public Builder name(String name) {
+            this.name = name;
+            return this;
+        }
+
+        public Builder sender(String snder) {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java]: s/snder/sender

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            this.sender = snder;
+            return this;
+        }
+
+        public Builder recipients(Collection<MailAddress> recipients) {
+            this.recipients = recipients.stream()
+                    .map(MailAddress::asPrettyString)
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java]: for sender you do the asPrettyString outside, and for recipients you do the opposite. We should choose the right way to do it.

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static MailQueueItemDTO from(ManageableMailQueue.MailQueueItemView mailQueueItemView) throws MailQueueException {
+        return builder()
+                .name(mailQueueItemView.getMail().getName())
+                .sender(mailQueueItemView.getMail().getSender().asPrettyString())
+                .recipients(mailQueueItemView.getMail().getRecipients())
+                .delayed(isDelayed(mailQueueItemView))
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java]: I would rather give the next delivery date

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 4]
```diff
@@ -140,4 +146,50 @@ private MailQueueDTO toDTO(ManageableMailQueue queue) {
...
+                    .message(String.format("%s can not be found", mailQueueName))
+                    .statusCode(HttpStatus.NOT_FOUND_404)
+                    .type(ErrorResponder.ErrorType.NOT_FOUND)
+                    .haltError());
+    }
+
+    private List<MailQueueItemDTO> listMessages(ManageableMailQueue queue) {
+        try {
+            return Iterators.toStream(queue.browse())
+                    .map(Throwing.function(MailQueueItemDTO::from))
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: if you don't sneakyThrow, you won't reach the catch block.

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 5]
```diff
@@ -140,4 +146,50 @@ private MailQueueDTO toDTO(ManageableMailQueue queue) {
...
+                    .collect(Guavate.toImmutableList());
+        } catch (MailQueueException e) {
+            LOGGER.info("Invalid request for getting the mail queue " + queue, e);
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: are you sure it's the only case you reach this block ?

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 6]
```diff
@@ -162,4 +167,52 @@ public void getMailQueueShouldReturnNotFoundWhenMailQueueDoesntExist() {
             .statusCode(HttpStatus.NOT_FOUND_404);
     }
 
+    @Test
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: we are missing error cases IMO

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 7]
```diff
@@ -150,7 +154,8 @@ private MailQueueDTO toDTO(ManageableMailQueue queue) {
     @GET
     @Path("/{mailQueueName}/messages")
     @ApiImplicitParams({
-        @ApiImplicitParam(required = true, dataType = "string", name = "mailQueueName", paramType = "path")
+        @ApiImplicitParam(required = true, dataType = "string", name = "mailQueueName", paramType = "path"),
+        @ApiImplicitParam(required = false, dataType = "boolean", name = DELAYED_QUERY_PARAM, paramType = "query")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: it lacks semantic explanation IMO

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 8]
```diff
@@ -155,7 +158,8 @@ private MailQueueDTO toDTO(ManageableMailQueue queue) {
     @Path("/{mailQueueName}/messages")
     @ApiImplicitParams({
         @ApiImplicitParam(required = true, dataType = "string", name = "mailQueueName", paramType = "path"),
-        @ApiImplicitParam(required = false, dataType = "boolean", name = DELAYED_QUERY_PARAM, paramType = "query")
+        @ApiImplicitParam(required = false, dataType = "boolean", name = DELAYED_QUERY_PARAM, paramType = "query"),
+        @ApiImplicitParam(required = false, dataType = "long", name = LIMIT_QUERY_PARAM, paramType = "query")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: we could document the default value here

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 9]
```diff
@@ -188,9 +192,16 @@ public void listMessages(Service service) {
                 .map(Boolean::parseBoolean);
     }
 
-    private List<MailQueueItemDTO> listMessages(ManageableMailQueue queue, Optional<Boolean> isDelayed) {
+    @VisibleForTesting long limit(String limitAsString) {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: you should reuse the Limit object and handle the 0 or less values

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java (hunk 2 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java]: I prefer `asString`

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+                .build();
+    }
+
+    private static boolean isDelayed(ManageableMailQueue.MailQueueItemView mailQueueItemView) {
+        return mailQueueItemView.getNextDelivery() != -1;
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/dto/MailQueueItemDTO.java]: magic value?

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 11]
```diff
@@ -162,4 +167,52 @@ public void getMailQueueShouldReturnNotFoundWhenMailQueueDoesntExist() {
...
+        when()
+            .get(FIRST_QUEUE + "/messages")
+        .then()
+            .statusCode(HttpStatus.OK_200)
+            .contentType(ContentType.JSON)
+            .body(".", hasSize(2));
+    }
+
+    @Test
+    public void listMessagesShouldReturnMessageDetailWhenSome() throws Exception {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: s/Detail/Details/ ?

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 12]
```diff
@@ -215,4 +217,53 @@ public void listMessagesShouldReturnMessageDetailWhenSome() throws Exception {
...
+            .get(FIRST_QUEUE + "/messages")
+        .then()
+            .statusCode(HttpStatus.OK_200)
+            .contentType(ContentType.JSON)
+            .body(".", hasSize(1));
+    }
+
+    @Ignore("MemoryMailQueueFactory doesn't support delay")
+    @Test
+    public void listMessagesShouldReturnDelayedMessagesAndAskFor() throws Exception {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: s/And/When/ ?

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 13]
```diff
@@ -266,4 +266,22 @@ public void listMessagesShouldReturnDelayedMessagesAndAskFor() throws Exception
             .contentType(ContentType.JSON)
             .body(".", hasSize(1));
     }
+
+    @Test
+    public void listMessagesShouldReturnOneMessageWhenMessagesAndAskForALimitOfOne() throws Exception {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: missing test cases: 0, invalid, <0, >1.

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesUnitTest.java [hunk 14]
```diff
@@ -62,4 +62,22 @@ public void isDelayedShouldBeEqualsToFalseWhenOtherValue() {
...
     }
+
+    @Test
+    public void limitShouldBeEqualsToDefaultValueWhenNull() {
+        long limit = testee.limit(null);
+        assertThat(limit).isEqualTo(MailQueueRoutes.DEFAULT_LIMIT_VALUE);
+    }
+
+    @Test
+    public void limitShouldBeToTheValueWhenOne() {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesUnitTest.java]: s/Be/BeEquals

**Comment by aduprat** [comment]
> static import for RestAssured, please
:+1:

**Comment by aduprat** [comment]
> What happens with a zero limit? With an invalid limit?
:+1:

**Comment by aduprat** [comment]
`push -f` to get rid of conflicts

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**